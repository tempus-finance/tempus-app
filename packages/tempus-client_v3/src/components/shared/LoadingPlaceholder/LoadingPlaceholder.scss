@use "sass:map";

$placeholder-widths: (
  'tiny': 43px,
  'small': 86px,
  'medium': 128px,
  'large': 182px,
);

$placeholder-heights: (
  'small': 12px,
  'medium': 24px,
);

$spacing: 8px;
$animation-duration: 2s;

.tc__loading-placeholder {
  display: flex;

  @each $size, $width in $placeholder-widths {
    &.tc__loading-placeholder__width-#{$size} {
      .tc__loading-placeholder__bar {
        width: $width;
      }
    }
  }

  @each $size, $height in $placeholder-heights {
    &.tc__loading-placeholder__height-#{$size} {
      .tc__loading-placeholder__circle {
        width: $height;
        animation: shine-loading-placeholder-#{$size}-circle $animation-duration infinite linear;
      }

      .tc__loading-placeholder__circle,
      .tc__loading-placeholder__bar {
        height: $height;
        border-radius: $height / 2;
      }
    }
  }

  .tc__loading-placeholder__circle {
    margin-right: $spacing;
  }

  .tc__loading-placeholder__bar {
    animation: shine-loading-placeholder-bar $animation-duration infinite linear;
  }

  .tc__loading-placeholder__circle,
  .tc__loading-placeholder__bar {
    background: linear-gradient(var(--loadingPlaceholderOverlay), var(--loadingPlaceholderOverlay)),
      linear-gradient(
        120deg,
        var(--loadingPlaceholderBase) 0%,
        var(--loadingPlaceholderBase) 5%,
        var(--loadingPlaceholderShine) 50%,
        var(--loadingPlaceholderBase) 95%,
        var(--loadingPlaceholderBase) 100%
      );
    background-size: 2 * map.get($placeholder-widths, 'large');
  }
}

@keyframes shine-loading-placeholder-bar {
  0% {
    background-position: -1 * map.get($placeholder-widths, 'large');
  }

  100% {
    background-position: map.get($placeholder-widths, 'large');
  }
}

@each $size, $height in $placeholder-heights {
  @keyframes shine-loading-placeholder-#{$size}-circle {
    0% {
      background-position: -1 * map.get($placeholder-widths, 'large') + $height + $spacing;
    }

    100% {
      background-position: map.get($placeholder-widths, 'large') + $height + $spacing;
    }
  }
}
