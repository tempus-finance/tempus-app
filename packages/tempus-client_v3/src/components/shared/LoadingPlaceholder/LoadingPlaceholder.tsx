import { FC } from 'react';
import './LoadingPlaceholder.scss';

export type LoadingPlaceholderWidth = 'tiny' | 'small' | 'medium' | 'large';
export type LoadingPlaceholderHeight = 'small' | 'medium';

export interface LoadingPlaceholderProps {
  width: LoadingPlaceholderWidth;
  height: LoadingPlaceholderHeight;
}

const LoadingPlaceholder: FC<LoadingPlaceholderProps> = props => {
  const { width, height } = props;

  return (
    <div
      // eslint-disable-next-line max-len
      className={`tc__loading-placeholder tc__loading-placeholder__width-${width} tc__loading-placeholder__height-${height}`}
    >
      <div className="tc__loading-placeholder__circle" />
      <div className="tc__loading-placeholder__bar" />
    </div>
  );
};

export default LoadingPlaceholder;
