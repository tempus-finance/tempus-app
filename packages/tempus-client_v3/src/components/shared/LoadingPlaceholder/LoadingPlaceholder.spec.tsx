import { render } from '@testing-library/react';
import LoadingPlaceholder, {
  LoadingPlaceholderHeight,
  LoadingPlaceholderProps,
  LoadingPlaceholderWidth,
} from './LoadingPlaceholder';

const subject = (props: LoadingPlaceholderProps) => render(<LoadingPlaceholder {...props} />);

describe('LoadingPlaceholder', () => {
  ['tiny', 'small', 'medium', 'large'].forEach(width => {
    ['small', 'medium'].forEach(height => {
      it(`renders a loading placeholder with ${width} width and ${height} height`, () => {
        const { container } = subject({
          width: width as LoadingPlaceholderWidth,
          height: height as LoadingPlaceholderHeight,
        });

        expect(container).not.toBeNull();
        expect(container).toMatchSnapshot();
      });
    });
  });
});
