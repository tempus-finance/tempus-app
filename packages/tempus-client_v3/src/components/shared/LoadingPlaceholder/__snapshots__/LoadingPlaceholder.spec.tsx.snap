// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingPlaceholder renders a loading placeholder with large width and medium height 1`] = `
<div>
  <div
    class="tc__loading-placeholder tc__loading-placeholder__width-large tc__loading-placeholder__height-medium"
  >
    <div
      class="tc__loading-placeholder__circle"
    />
    <div
      class="tc__loading-placeholder__bar"
    />
  </div>
</div>
`;

exports[`LoadingPlaceholder renders a loading placeholder with large width and small height 1`] = `
<div>
  <div
    class="tc__loading-placeholder tc__loading-placeholder__width-large tc__loading-placeholder__height-small"
  >
    <div
      class="tc__loading-placeholder__circle"
    />
    <div
      class="tc__loading-placeholder__bar"
    />
  </div>
</div>
`;

exports[`LoadingPlaceholder renders a loading placeholder with medium width and medium height 1`] = `
<div>
  <div
    class="tc__loading-placeholder tc__loading-placeholder__width-medium tc__loading-placeholder__height-medium"
  >
    <div
      class="tc__loading-placeholder__circle"
    />
    <div
      class="tc__loading-placeholder__bar"
    />
  </div>
</div>
`;

exports[`LoadingPlaceholder renders a loading placeholder with medium width and small height 1`] = `
<div>
  <div
    class="tc__loading-placeholder tc__loading-placeholder__width-medium tc__loading-placeholder__height-small"
  >
    <div
      class="tc__loading-placeholder__circle"
    />
    <div
      class="tc__loading-placeholder__bar"
    />
  </div>
</div>
`;

exports[`LoadingPlaceholder renders a loading placeholder with small width and medium height 1`] = `
<div>
  <div
    class="tc__loading-placeholder tc__loading-placeholder__width-small tc__loading-placeholder__height-medium"
  >
    <div
      class="tc__loading-placeholder__circle"
    />
    <div
      class="tc__loading-placeholder__bar"
    />
  </div>
</div>
`;

exports[`LoadingPlaceholder renders a loading placeholder with small width and small height 1`] = `
<div>
  <div
    class="tc__loading-placeholder tc__loading-placeholder__width-small tc__loading-placeholder__height-small"
  >
    <div
      class="tc__loading-placeholder__circle"
    />
    <div
      class="tc__loading-placeholder__bar"
    />
  </div>
</div>
`;

exports[`LoadingPlaceholder renders a loading placeholder with tiny width and medium height 1`] = `
<div>
  <div
    class="tc__loading-placeholder tc__loading-placeholder__width-tiny tc__loading-placeholder__height-medium"
  >
    <div
      class="tc__loading-placeholder__circle"
    />
    <div
      class="tc__loading-placeholder__bar"
    />
  </div>
</div>
`;

exports[`LoadingPlaceholder renders a loading placeholder with tiny width and small height 1`] = `
<div>
  <div
    class="tc__loading-placeholder tc__loading-placeholder__width-tiny tc__loading-placeholder__height-small"
  >
    <div
      class="tc__loading-placeholder__circle"
    />
    <div
      class="tc__loading-placeholder__bar"
    />
  </div>
</div>
`;
